src/crypto.ts

```diff
@@ -0,0 +1,361 @@
+ * An interface describing a cryptographic provider.
+ */
+export interface CryptoProvider {
+	getHash(algorithm: string): HashFunction;
+	getSign(algorithm: string): SignFunction;
+}
+
+/**
+ * The current provider. Providers supply concrete implementations of the API described here. Users should not typically
+ * need to access providers directly.
+ */
+let provider: Promise<CryptoProvider>;
+
+/**
+ * Gets the HashFunction for a particular algorithm. The algorithm is specified as a string for simplicity and
+ * extensibility.
+ */
+export function getHash(algorithm: string): HashFunction {
+	let realHash: HashFunction;
+	const hashPromise = new Promise<HashFunction>(function (resolve, reject) {
+		getProvider().then(
+			function (provider) {
+				try {
+					realHash = provider.getHash(algorithm);
+					resolve(realHash);
+				}
+				catch (error) {
+					reject(error);
+				}
+			},
+			function (error) {
+				reject(error);
+			}
+		);
+	});
+
+	// Return a wrapper that will defer calls to the hash until a provider has been loaded.
+	const hashFunction = <HashFunction> function (data: Data, codec?: Codec): Promise<Binary> {
+		if (realHash) {
+			return realHash(<any> data, codec);
+		}
+		return hashPromise.then<Binary>(function (hash) {
+			return hash(<any> data, codec);
+		});
+	}
+
+	// Return a wrapper class that will defer calls until a provider has been loaded and an actual Hasher instance has
+	// been created.
+	hashFunction.create = function<T extends Data> (codec?: Codec): Hasher<T> {
+		return new HasherWrapper(hashPromise, codec);
+	};
+
+	return hashFunction;
+}
+
+/**
+ * Gets the SignFunction for a particular algorithm. The algorithm is specified as a string for simplicity and
+ * extensibility.
+ */
+export function getSign(algorithm: string): SignFunction {
+	let realSign: SignFunction;
+	const signPromise = new Promise<SignFunction>(function (resolve, reject) {
+		getProvider().then(
+			function (provider) {
+				try {
+					realSign = provider.getSign(algorithm);
+					resolve(realSign);
+				}
+				catch (error) {
+					reject(error);
+				}
+			},
+			function (error) {
+				reject(error);
+			}
+		);
+	});
+
+	// Return a wrapper that will defer calls to the hash until a provider has been loaded.
+	const signFunction = <SignFunction> function (key: Key, data: Data, codec?: Codec): Promise<Binary> {
+		if (realSign) {
+			return realSign(key, <any> data, codec);
+		}
+		return signPromise.then<Binary>(function (sign) {
+			return sign(key, <any> data, codec);
+		});
+	}
+
+	// Return a wrapper class that will defer calls until a provider has been loaded and an actual Hasher instance has
+	// been created.
+	signFunction.create = function<T extends Data> (key: Key, codec?: Codec): Signer<T> {
+		return new SignerWrapper(signPromise, key, codec);
+	};
+
+	return signFunction;
+}
+
+/**
+ * Returns a promise that resolves to the current provider object.
+ */
+export function getProvider(): Promise<CryptoProvider> {
```

Comment:
It was meant for informative purposes, so you could check which provider is actually in use, not to give you a handle to something to use directly. We can hide it for now and export later if desired.